### ISSUE TITLE
zsync_curl cannot download from Bintray

### ISSUE THREAD

**Issue by probonopd** [probonopd/zsync-curl#1]
I cannot get `zsync_curl` to work with `.zsync` files hosted on [Bintray](https://bintray.com/).

Using zsync would be very advantageous since with it, users would have to download only the portion of a file that has changed since the last version, rather than the entire file. Ubuntu, for example, is using zsync for their nightly ISO builds. Since each [AppImage](http://appimage.org/) is also an ISO, I want to use zsync too.

To debug this, I built a version of `zsync_curl` with `CURLOPT_VERBOSE` set to 1 like this:

```
# Fedora 23
sudo yum -y install libcurl-devel gcc
git clone https://github.com/probonopd/zsync-curl.git
cd zsync-curl/
sed -i -e 's/CURLOPT_VERBOSE, 0L/CURLOPT_VERBOSE, 1L/g' http.c # Patch verbosity
./configure && make && sudo make install
```

Now we can see what is going on: [log.txt](https://github.com/probonopd/zsync-curl/files/60682/log.txt)

It looks like `Range: bytes=0-40959,98304-135303167` results in `HTTP/1.1 403 Forbidden`.

Well, on [this](https://www.jfrog.com/confluence/display/RTF/Artifactory+3.7) page we can read:

> Range Requests
> 
> Artifactory now supports HTTP Range Requests which enable smart and efficient multi-threaded download of large files by letting you specify a range of bytes that should be downloaded when files are streamed to you.

So the question is: Why are we getting `HTTP/1.1 403 Forbidden` from [CloudFront](https://aws.amazon.com/de/cloudfront/) and can we change anything in zsync-curl so that this does not happen?

Looks like it deduces from `Content-Disposition: attachment;filename="Scribus-1.5.1.svn.20620-x86_64.AppImage.zsync` that we are `downloading from https://d29vzk4ow07wi7.cloudfront.net/Scribus-1.5.1.svn.20620-x86_64.AppImage` but that is not correct, the full URL is much longer as can be seen from the GET request above it.

Here is some working `bash` code that shows that partial download is possible from CloudFront, and shows where I suspect the bug:

```
URL=http://dl.bintray.com/probono/AppImages/Scribus-1.5.1.svn.20620-x86_64.AppImage.zsync

# Get the real URL of the zsync file from the redirect
URLREDIRECTED=$(curl -v "$URL" 2>&1  | grep "< Location" | cut -d " " -f 3 | xargs | strings)

# Get the URL from inside the zsync file
INSIDEPATH=$(curl -v --header "Range: bytes=0-$((256))" "$URLREDIRECTED" | strings | grep "URL: " | cut -d " " -f 2)

# Now the magic must happen: We MUST NOT interpret the URL 
# from inside the zsync file relative to the redirected URL
# but instead we must interpret it relative to the original URL from step 1

if [ ! $(echo $INSIDEPATH | grep -e "^http") ] ; then
  ISOURL="$URL/../$INSIDEPATH" ##### I suspect this is where things go sour
else
  ISOURL="$INSIDEPATH"
fi

ISOURLREDIRECTED=$(curl -v "$ISOURL" 2>&1  | grep "< Location" | cut -d " " -f 3 | xargs | strings)

# Let's get the ISO header from within the file
curl -v --header "Range: bytes=$((16*2048))-$((16*2048+2048))" "$ISOURLREDIRECTED" | hexdump -C
```

If we can read a string saying `ISOIMAGE` here, then partial download works.

Putting the absolute URL into the `.zsync` file gets us to the next problem:

```
* Connected to dl.bintray.com (5.153.24.114) port 80 (#0)
> GET /probono/AppImages/Scribus-1.5.1.svn.20620-x86_64.AppImage HTTP/1.1
Host: dl.bintray.com
Range: bytes=0-40959,98304-135303167
Accept: */*
```

This is of course bogus. The `Range:` header should be sent to the redirected URL, not the original URL.

Overall, the whole logic of how to handle redirects is completely screwed up. It should be like this:
- From the `.zsync` file, get the file path to the ISO. If it is a relative path, then do **NOT** append it to the redirected URL but to the original download URL of the `.zsync` file
- Once we have the URL of the ISO, get its **redirected URL**. Use that for all requests involving `Range:` headers

Looks like this has been [discussed in 2008](http://sourceforge.net/p/zsync/mailman/message/18830652/):

> It has been suggested that zsync could hit the redirect once and remember the 
> new URL for the rest of the session, which would be easy enough to implement. 
> But, while this would be valid for 301 redirects, it seems to directly 
> violate the intention of 302 redirects, where the client is meant to 
> re-request the original URL every time. Then again, the spec wasn't written 
> with this scenario in mind.
> 
> That said, since I last thought about it, I haven't had any better idea, so I 
> guess I should go ahead, ignore the spec, and simply treat all redirects as 
> permanent-for-this-session.

Some redirect handling was added, but it does not work because it does not do the following:
- For each URL, check _once_ whether it is redirected (and if it is, then use the redirected URL for subsequent requests)
- When calculating the ISO URL from a relative path given in the `.zsync` file, then do _not_ calculate it relative to the redirected URL of the `.zync` file, but calculate it relative to the original (pre-redirect) URL. Using this newly calculated URL, check if it is redirected somewhere. If it is, use the resulting redirected URL for any subsequent requests, especially those involving `Range:` headers

In other words: For each URL it should be checked exactly **once** if it is redirected somewhere, and if it is, the redirected URL should be used. However it must **NOT** be assumed that if the `.zsync` URL is redirected somewhere, then the ISO URL is redirected to the same place (as with CloudFront, the two URLs get redirected to different locations).

**Comment by probonopd** [comment]
Here is some working `bash` code that shows that partial download is possible from CloudFront, and shows where I suspect the bug:

```
URL=http://dl.bintray.com/probono/AppImages/Scribus-1.5.1.svn.20620-x86_64.AppImage.zsync

# Get the real URL of the zsync file from the redirect
URLREDIRECTED=$(curl -v "$URL" 2>&1  | grep "< Location" | cut -d " " -f 3 | xargs | strings)

# Get the URL from inside the zsync file
INSIDEPATH=$(curl -v --header "Range: bytes=0-$((256))" "$URLREDIRECTED" | strings | grep "URL: " | cut -d " " -f 2)

# Now the magic must happen: We MUST NOT interpret the URL 
# from inside the zsync file relative to the redirected URL
# but instead we must interpret it relative to the original URL from step 1

ISOURL="$URL/../$INSIDEPATH" ##### I suspect this is where things go sour

ISOURLREDIRECTED=$(curl -v "$ISOURL" 2>&1  | grep "< Location" | cut -d " " -f 3 | xargs | strings)

# Let's get the ISO header from within the file
curl -v --header "Range: bytes=$((16*2048))-$((16*2048+2048))" "$ISOURLREDIRECTED" | hexdump -C
```

If we can read a string saying `ISOIMAGE` here, then partial download works.

**Comment by probonopd** [comment]
Putting the absolute URL into the `.zsync` file not gets us the next problem:

```
* Connected to dl.bintray.com (5.153.24.114) port 80 (#0)
> GET /probono/AppImages/Scribus-1.5.1.svn.20620-x86_64.AppImage HTTP/1.1
Host: dl.bintray.com
Range: bytes=0-40959,98304-135303167
Accept: */*
```

This is of course bogus. The `Range:` header should be sent to the redirected URL, not the original URL.

Overall, the whole logic of how to handle redirects is completely screwed up. It should be like this:

 * From the `.zsync` file, get the file path to the ISO. If it is a relative path, then do __NOT__ append it to the redirected URL but to the original download URL of the `.zsync` file
 * Once we have the URL of the ISO, get its __redirected URL__. Use that for all requests involving `Range:` headers

**Comment by probonopd** [comment]
The saga continues: Turns out that the server responds differently to HEAD and GET requests, so we must not use HEAD requests to find the final payload URL to download pieces from.

```
<probono> note that I _don't_ want to download any content
<probono> just find out the final redirected URL
<dreisner> then why not set CURLOPT_NOBODY and CURLOPT_FOLLOWLOCATION ?
<dreisner> and then use CURLINFO_EFFECTIVE_URL
<dreisner> otherwise, you'll have to follow the redirect chain yourself.
<probono> I am trying what you describe above
<probono> when I do that, the first "302 Found" is observed but the second redirect, "302" is not
<probono> I checked with wget:
<probono> wget https://bintray.com/artifact/download/probono/AppImages/Leafpad-0.8.17-x86_64.AppImage
(...)
<dreisner> wget shows me 1 redirect.
<dreisner> wget --spider https://bintray.com/artifact/download/probono/AppImages/Leafpad-0.8.17-x86_64.AppImage
<dreisner> oh, so your server responds differently to HEAD and GET requests
<dreisner> that's nasty.
<probono> indeed. when you invoke wget without spider mode then you see what I mean
(...)
<dreisner> then you need to fix your server
<dreisner> make it respond the same way to HEAD and GET requests
(...)
<probono> I could make every range request to the non-final URL, but that would mean many extraneous redirects
<dreisner> that seems like a tiny fraction of the overall cost.
(...)
<probono> use the C equivalent of 
<probono> curl -v -L --range 0-0 -o /dev/null https://bintray.com/artifact/download/probono/AppImages/Leafpad-0.8.17-x86_64.AppImage
<probono> to find out the final URL
```

**Comment by probonopd** [comment]
Fixed in https://github.com/probonopd/zsync-curl/commit/c089100ec44c1be231593fa31dfb99ee2727f9f6
